lib/zzhc/zzhc.c

```diff
@@ -0,0 +1,600 @@
+		if (rc < 0) {
+			return;
+		}
+
+		LOG_DBG("+CEREG: %d", rc);
+		if (rc == 1 || rc == 5) {
+			LOG_DBG("Registered to operator network");
+			ctx->registered = true;
+			sem_post(ctx->sem);
+		} else {
+			LOG_DBG("De-registered from operator network");
+			ctx->registered = false;
+		}
+	}
+}
+
+/**@brief Search and replace. */
+static void search_char_and_replace(char *data, char tok, char substitute)
+{
+	char *cptr = data;
+
+	while (cptr != NULL) {
+		cptr = strchr(cptr, tok);
+		if (cptr != NULL) {
+			*cptr++ = substitute;
+		}
+	}
+}
+
+/**@brief Build encrypted json file. */
+static int build_json(struct zzhc *ctx, char *buff, int len)
+{
+	char l_buff[JSON_MAX_LEN];
+	char duple[96];
+	char *cptr;
+	int i = 0, rc;
+
+	l_buff[0] = 0;
+
+	/* Self-registration protocol version. */
+	sprintf(duple, "{\"REGVER\":\"%d\",", REGVER);
+	strcat(l_buff, duple);
+
+	/* IMEI */
+	if (at_cmd_xfer("AT+CGSN=1", ctx->at_resp, AT_RESPONSE_LEN) != 0) {
+		return -EIO;
+	}
+	cptr = strchr(ctx->at_resp, '"');
+	if (cptr == NULL) {
+		return -EILSEQ;
+	}
+	cptr++;
+	cptr[IMEI_LEN] = 0;
+
+	sprintf(duple, "\"MEID\":\"%s\",", cptr);
+	strcat(l_buff, duple);
+
+	/* MODEL */
+	if (at_cmd_xfer("AT+CGMM", ctx->at_resp, AT_RESPONSE_LEN) != 0) {
+		return -EIO;
+	}
+	search_char_and_replace(ctx->at_resp, '-', '_');
+	cptr = strstr(ctx->at_resp, "\r\n");
+	if (cptr != NULL) {
+		*cptr = 0;
+	}
+
+	rc = read_chip_rev(&i);
+	if (rc != 0) {
+		return rc;
+	}
+	LOG_DBG("i=%d", i);
+
+	sprintf(duple, "\"MODEL\":\"%s-%s_B%cA\",", MANUFACTURER_CODE,
+		ctx->at_resp, chip_rev[i]);
+	strcat(l_buff, duple);
+
+	/* SWVER */
+	if (at_cmd_xfer("AT+CGMR", ctx->at_resp, AT_RESPONSE_LEN) != 0) {
+		return -EIO;
+	}
+	cptr = strstr(ctx->at_resp, "\r\n");
+	if (cptr != NULL) {
+		*cptr = 0;
+	}
+	sprintf(duple, "\"SWVER\":\"%s\",", ctx->at_resp);
+	strcat(l_buff, duple);
+
+	/* SIM1ICCID */
+	sprintf(duple, "\"SIM1ICCID\":\"%s\",", ctx->ueiccid);
+	strcat(l_buff, duple);
+
+	/* SIM1LTEIMSI */
+	if (at_cmd_xfer("AT+CIMI", ctx->at_resp, AT_RESPONSE_LEN) != 0) {
+		return -EIO;
+	}
+	cptr = strstr(ctx->at_resp, "\r\n");
+	if (cptr != NULL) {
+		*cptr = 0;
+	}
+	sprintf(duple, "\"SIM1LTEIMSI\":\"%s\"}", ctx->at_resp);
+	strcat(l_buff, duple);
+
+	return base64(buff, len, l_buff, strlen(l_buff));
+}
+
+/**@brief Compile HTTP request packet. */
+static void build_http_req(struct zzhc *ctx)
+{
+	char buff[64];
+	char json_buff[JSON_MAX_LEN];
+	int rc;
+
+	strcat(ctx->http_pkt, "POST / HTTP/1.1\r\n");
+
+	sprintf(buff, "Host: %s:%d\r\n", hostname, SERVER_PORT);
+	strcat(ctx->http_pkt, buff);
+
+	strcat(ctx->http_pkt, "User-Agent: zzhc/1.0.0\r\n");
+	strcat(ctx->http_pkt, "Accept: */*\r\n");
+	strcat(ctx->http_pkt, "x-forwarded-for: 117.61.5.70\r\n");
+	strcat(ctx->http_pkt, "Content-Type: application/encrypted-json\r\n");
+	strcat(ctx->http_pkt, "Connection: close\r\n");
+	strcat(ctx->http_pkt, "Content-Length: ");
+
+	rc = build_json(ctx, json_buff, JSON_MAX_LEN);
```

Comment:
In the push on 5th Nov, 2019, the return code from `build_json()` is propagated to `build_http_req()`, and then to `fsm()`. In case of errors, it will result in a state change to `STATE_RETRY_TIMEOUT` for further retries.